alphabase/psm_reader/sage_reader.py

```diff
@@ -66,90 +259,242 @@ def lookup_modification(
+            mod_site = (
+                "-1"
+                if matched_mod.iloc[0]["is_cterm"]
+                else str(m.start() - accumulated_non_sequence_chars)
+            )
+            mod_tag_len = m.end() - m.start() + 1
+
+        else:
+            # side chain mod
+            matched_mod = mod_translation_df[
+                (mod_translation_df["modification"] == group)
+                & (mod_translation_df["previous_aa"] == previous_char)
+            ]
+            if len(matched_mod) == 0:
+                return None, None
+
+            mod_tag_len = m.end() - m.start()
+            matched_mod_name = matched_mod.iloc[0]["matched_mod_name"]
+            mod_site = str(m.start() - accumulated_non_sequence_chars)
 
-    site_list = []
-    mod_list = []
+        accumulated_non_sequence_chars += mod_tag_len
+        mod_sites.append(mod_site)
+        mod_names.append(matched_mod_name)
 
-    error = False
+    return ";".join(mod_sites), ";".join(mod_names)
 
-    match_delta = 0
 
-    for match in matches:
-        match_start, match_end = match.start(), match.end()
-        previous_aa = sequence[match_start - 1] if match_start > 0 else "Any_N-term"
-        mass_observed = float(match.group(2)) * (1 if match.group(1) == "+" else -1)
+def apply_translate_modifications(
+    df: pd.DataFrame, mod_translation_df: pd.DataFrame
+) -> pd.DataFrame:
+    """Apply the translation of modifications to the PSMs.
 
-        mod = lookup_modification(
-            mass_observed, previous_aa, mod_annotated_df, ppm_tolerance=ppm_tolerance
+    Parameters
+    ----------
+
+    df : pd.DataFrame
+        The PSM dataframe with column 'modified_sequence'.
+
+    mod_translation_df : pd.DataFrame
+        The annotated modification dataframe.
+
+    Returns
+    -------
+
+    pd.DataFrame
+        The PSM dataframe with columns 'mod_sites' and 'mods'.
+
+    """
+
+    df["mod_sites"], df["mods"] = zip(
+        *df["modified_sequence"].apply(
+            lambda x: translate_modifications(x, mod_translation_df)
         )
-        if mod is not None:
-            site_list.append(str(match_start - match_delta))
-            mod_list.append(mod)
+    )
+    return df
 
-        else:
-            error = True
-            print(
-                f"No modification found for mass {mass_observed} at position {match_start} with previous aa {previous_aa}"
+
+def _batchify_df(df: pd.DataFrame, mp_batch_size: int) -> typing.Generator:
+    """Internal funciton for applying translation modifications in parallel.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The PSM dataframe.
+
+    mp_batch_size : int
+        The batch size for parallel processing.
+
+    Returns
+    -------
+
+    typing.Generator
+        A generator for the batchified dataframe.
+    """
+
+    for i in range(0, len(df), mp_batch_size):
+        yield df.iloc[i : i + mp_batch_size, :]
```

Comment:
We have to get the claude bot :D
```
The function _batchify_df as written will return the last batch correctly, but there's an important detail to consider:

The function uses a range with a step size of mp_batch_size, which means it will generate batches of the specified size.
The slicing operation df.iloc[i : i + mp_batch_size, :] will automatically adjust if the end index (i + mp_batch_size) is greater than the length of the dataframe.
This means that the last batch will contain the remaining rows, even if it's smaller than mp_batch_size.

For example, if the dataframe has 105 rows and mp_batch_size is 30:

The first three batches will have 30 rows each
The last batch will have the remaining 15 rows

So yes, this function will return the last batch correctly, regardless of whether it's a full batch or a partial one. It's a well-implemented batching function that handles the edge case of the final, potentially smaller batch appropriately.
If you want to make this behavior more explicit, you could add a comment in the function docstring mentioning that the last batch may be smaller than mp_batch_size. But functionally, the code is correct as it is.
```

Reply:
I feel we are one step before a new "let me google that for you" .. next time the prompt will do :-D

P.S.: congrats on your "well-implemented batching function", the paid AIs are really charming :-D